### DISCUSSION TITLE
grub boot manager broken - how to install zfsbotomenu

### DISCUSSION THREAD

**Discussion by SimonBard** [zbm-dev/zfsbootmenu#600] [Questions]
My system with grub does not want to start. Most probably because I have done a snapshot of the boot partition ([see here)](https://github.com/openzfs/zfs/issues/13873). 

Can I somehow install zfsbootmenu? How can I perform this? I guess it will not help to start a live usb system and perform the steps [here](https://github.com/zbm-dev/zfsbootmenu/wiki/Complete-beginner's-guide-to-installing-ZFSBootMenu#12-copy-zfsbootmenu-onto-the-efi-partitions).

**Comment by ahesford** [comment]
The easiest way is to start with a [portable installation](https://docs.zfsbootmenu.org/en/v2.3.x/general/portable.html) that will let you boot from a USB drive. This should give you all the tools you need to migrate `/boot` into the root filesystem in such a way that ZFSBootMenu will recognize your boot environment. Af first, ZFSBootMenu will probably fail to recognize any boot environments and try dropping you into a recovery shell. Assuming your root filesystem is `rpool/ROOT/debian` and your boot filesystem is `bpool/ROOT/debian`, you should be able to:
1. Confirm that `bpool` and `rpool` are both imported. If not, you can manually import each pool, or you can try running `/libexec/zfunc import_pool` which will try to import both.
2. Make sure the pools are writable:
    ```
    set_rw_pool rpool
    set_rw_pool bpool
    ```
4. Use our `mount_zfs` helper to mount both the boot and root filesystems:
    ```
    mount_zfs bpool/ROOT/debian
    allow_rw=yes mount_zfs rpool/ROOT/debian
    ```
   each call will print the path where the filesystem is mounted, which (in this case) will be, respectively,
    ```
    /zfsbootmenu/environments/bpool/ROOT/debian/mnt
    /zfsbootmenu/environments/rpool/ROOT/debian/mnt
    ```
5. Copy the contents of the boot filesystem to the `boot` subdirectory of the root:
    ```
    cd /zfsbootmenu/environments/bpool/ROOT/debian/mnt
    mkdir -p /zfsbootmenu/environments/rpool/ROOT/debian/mnt/boot
    cp -a . /zfsbootmenu/environments/rpool/ROOT/debian/mnt/boot
    ````
6. Make sure the boot pool won't be mounted at next boot:
    ```
    zfs set canmount=noauto bpool/ROOT/debian
    ```
7. Exit the recover shell. You should now see your environment in the menu.

**Comment by SimonBard** [comment]
Many thanks for the quick response!

I am already failing with the portable version. The usb stick gets recognised, but grub starts. So zfsbootmenu is not starting from the usb stick. 

![Screenshot from 2024-03-05 09-07-58](https://github.com/zbm-dev/zfsbootmenu/assets/63670466/d2435f6d-7f7a-4525-8f49-bdc89bc88fb3)
![PXL_20240305_083230217](https://github.com/zbm-dev/zfsbootmenu/assets/63670466/405ca490-4dac-4ad7-b7c7-62898faf057a)

**Reply by classabbyamp** [reply]
select `UEFI: USB, Partition 1`, not `USB: USB`

**Reply by SimonBard** [reply]
Interesting, I set UEFI: USB, Partition 1 as the second option, but it would not load. Setting it at first position did the job. 

But I am still struggling, thanks for replying to quick and patiently.
Both pools are imported, but not mounted. I use ubuntu 22.04. I can also start from a live usb with ubuntu to get the filestructure if that helps. 

![PXL_20240305_102757299 1](https://github.com/zbm-dev/zfsbootmenu/assets/63670466/b6250375-cffb-41a2-b19e-9142dfa7bf4d)

**Reply by SimonBard** [reply]
Ok interesting, the directory for ubuntu is: 
bpool/BOOT/tzf9vc

but I still fail:

![PXL_20240305_110055974 2](https://github.com/zbm-dev/zfsbootmenu/assets/63670466/e93b8f0a-f98b-4188-8842-a1092c16a0ef)

I fail to umount so I can mount it again:
![PXL_20240305_111424212 1](https://github.com/zbm-dev/zfsbootmenu/assets/63670466/ea0bfb4e-2c1c-4b1a-926d-67b5fc5bf481)


![PXL_20240305_112616045 MP 1](https://github.com/zbm-dev/zfsbootmenu/assets/63670466/6449462c-91c6-4c73-b1c6-c513b4e22dd7)

**Reply by ahesford** [reply]
That's not a directory but a filesystem name, and you're mistyping the "_tz" as "_ty".

**Comment by SimonBard** [comment]
> That's not a directory but a filesystem name, and you're mistyping the "_tz" as "_ty".

Many thanks. It finally worked. 

The system now only starts when I have the usb stick plugged in. How can I avoid that?
Can I destroy the original bpool, create a new one and copy the data from the partition of my USB stick to the pool?

And can I define that the system it starting itself without having to press Enter?

**Reply by zdykstra** [reply]
The USB drive has an EFI System Partition - the contents can't be used on a ZFS dataset. Your options are to try to recreate your boot pool, datasets and contents or to remove zsys and stick with ZFSBootMenu.

If you stick with ZBM, you'll just need to hit MOD+D on the main screen to set a default boot entry.
